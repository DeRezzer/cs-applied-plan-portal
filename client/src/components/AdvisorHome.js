/** @jsx jsx */

import NavBar from "./Navbar";
import PageSpinner from "./general/PageSpinner";
import {useEffect, useState} from "react";
import {css, jsx} from "@emotion/core";
import {withRouter} from "react-router-dom";
import PropTypes from "prop-types";

function AdvisorHome(props) {

  const [loading, setLoading] = useState(true);
  const [plans, setPlans] = useState([]);
  const [errorMessage, setErrorMessage] = useState("");

  const style = css`

    #plan-data-container {
      margin: 0 auto;
      width: 50%;
    }

    #plan-selection-container {
      position: relative;
      top: 75px;
    }

    .home-error-message-container {
      position: relative;
      top: 100px;
    }
    
    .advisor-plans-table {
      position: relative;
      top: 100px;
    }

  `;

  useEffect(() => {
    fetchPlans();

  }, [props.history]);

  async function fetchPlans() {
    try {
      setErrorMessage("");
      setLoading(true);
      setPlans([]);

      const selectStatus = document.getElementById("select-status");
      const statusValue = selectStatus.options[selectStatus.selectedIndex].value;

      const selectTime = document.getElementById("select-time");
      const timeValue = selectTime.options[selectTime.selectedIndex].value;

      const selectOrder = document.getElementById("select-order");
      const orderValue = selectOrder.options[selectOrder.selectedIndex].value;

      const server = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;
      const getUrl = `http://${server}/plan/status/${statusValue}/${timeValue}/${orderValue}` +
        `?accessToken=${props.token}`;
      let obj = {};

      const results = await fetch(getUrl);
      setLoading(false);
      if (results.ok) {
        obj = await results.json();
        setPlans(obj.data);
      } else {
        // we got a bad status code. Show the error
        obj = await results.json();
        setErrorMessage(obj.error);
      }
    } catch (err) {
      // send to 500 page if a server error happens while fetching plan
      setErrorMessage("An internal server error occurred. Please try again later.");
    }
  }

  function renderStatus(status) {
    switch (status) {
      case 0:
        return "Rejected";
      case 1:
        return "Awaiting student changes";
      case 2:
        return "Awaiting review";
      case 3:
        return "Awaiting final review";
      case 4:
        return "Accepted";
      default:
        return "Undefined status";
    }
  }

  function goToPlan(plan) {
    window.location.href = `/viewPlan/${plan.planId}`;
  }

  return (
    <div css={style}>
      <PageSpinner loading={loading} />
      <NavBar showSearch={true} searchContent={"Search for plans"}/>
      <div id="plan-data-container">
        <div id="plan-selection-container">
          <select id="select-status" className="advisor-plan-select">
            <option value="5">Any</option>
            <option value="2">Awaiting Review</option>
            <option value="3">Awaiting final review</option>
            <option value="1">Awaiting student changes</option>
            <option value="4">Accepted</option>
            <option value="0">Rejected</option>
          </select>
          <select id="select-time" className="advisor-plan-select">
            <option value="1">Time Created</option>
            <option value="0">Time Updated</option>
          </select>
          <select id="select-order" className="advisor-plan-select">
            <option value="1">Ascending</option>
            <option value="0">Decending</option>
          </select>
          <button id="search-plan-status-button" onClick={() => { fetchPlans(); }}>
            Search
          </button>
        </div>
        <div className="home-error-message-container">{errorMessage}</div>
        <table className="advisor-plans-table">
          <tbody>
            <tr>
              <th className="student-plans-data">User Name</th>
              <th className="student-plans-data">User ID</th>
              <th className="student-plans-data">Plan Name</th>
              <th className="student-plans-data">Status</th>
              <th className="student-plans-data">Created</th>
              <th className="student-plans-data">Updated</th>
            </tr>
            {plans ? plans.map(plan =>
              <tr key={plan.planId} onClick={() => goToPlan(plan)}>
                <td className="student-plans-data" key={plan.planId + "a"}>
                  {plan.firstName + " " + plan.lastName}
                </td>
                <td className="student-plans-data" key={plan.planId + "b"}>{plan.userId}</td>
                <td className="student-plans-data" key={plan.planId + "c"}>{plan.planName}</td>
                <td className="student-plans-data" key={plan.planId + "d"}>{renderStatus(plan.status)}</td>
                <td className="student-plans-data" key={plan.planId + "e"}>{plan.created}</td>
                <td className="student-plans-data" key={plan.planId + "f"}>{plan.lastUpdated}</td>

              </tr>) : null}
          </tbody>
        </table>
      </div>
    </div>
  );
}
export default withRouter(AdvisorHome);

AdvisorHome.propTypes = {
  history: PropTypes.object,
  token: PropTypes.string
};