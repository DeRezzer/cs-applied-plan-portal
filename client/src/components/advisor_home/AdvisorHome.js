/** @jsx jsx */

import NavBar from "../Navbar";
import PageSpinner from "../general/PageSpinner";
import {useEffect, useState} from "react";
import {getToken} from "../../utils/authService";
import RecentPlans from "./RecentPlans";
import FindPlans from "./FindPlans";
import SearchResults from "./SearchResults";
import PageInternalError from "../general/PageInternalError";
import {css, jsx} from "@emotion/core";
import PropTypes from "prop-types";

function AdvisorHome() {

  const [loading, setLoading] = useState(false);
  const [pageError, setPageError] = useState(0);
  const [cursor, setCursor] = useState({
    primary: "null",
    secondary: "null"
  });
  const [recentPlans, setRecentPlans] = useState([]);
  const [plans, setPlans] = useState([]);
  const [errorMessage, setErrorMessage] = useState("");
  const [searchFields, setSearchFields] = useState({
    textValue: "*",
    searchValue: 0,
    statusValue: 0,
    sortValue: 0,
    orderValue: 0
  });

  const style = css`

    #advisor-home-container {
      position: absolute;
      top: 75px;
      margin: 0 auto;
      width: 100%;
    }

    #advisor-home-contents-container {
      margin: 25px auto;
      width: 50%;
    }

    .home-error-message-container {
      text-align: center;
      margin: 0 auto;
    }

  `;

  useEffect(() => {
    getRecentPlans();
  }, []);

  async function getRecentPlans() {
    try {
      setErrorMessage("");
      setLoading(true);

      const token = getToken();
      const server = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;
      const getUrl = `http://${server}/plan/recent/?accessToken=${token}`;
      let obj = {};

      const results = await fetch(getUrl);
      setLoading(false);
      if (results.ok) {
        obj = await results.json();
        setRecentPlans(obj.plans);
      } else {
        // we got a bad status code.
        obj = await results.json();
        if (results.status === 500) {
          setPageError(500);
        }
      }
    } catch (err) {
      // send to 500 page if a server error happens while fetching plan
      setPageError(500);
    }
  }

  async function searchPlans(cursor) {
    try {
      setErrorMessage("");
      setLoading(true);

      const text = document.getElementById("input-search");
      let textValue = text.value;
      if (textValue === "") {
        textValue = "*";
      }

      const search = document.getElementById("select-search");
      let searchValue = search.options[search.selectedIndex].value;

      const status = document.getElementById("select-status");
      let statusValue = status.options[status.selectedIndex].value;

      const sort = document.getElementById("select-sort");
      let sortValue = sort.options[sort.selectedIndex].value;

      const order = document.getElementById("select-order");
      let orderValue = order.options[order.selectedIndex].value;

      // only set the search values if we are performing a new search
      if (cursor.primary === "null") {
        setSearchFields({
          searchValue: searchValue,
          statusValue: statusValue,
          sortValue: sortValue,
          orderValue: orderValue
        });
      } else {
        searchValue = searchFields.searchValue;
        statusValue = searchFields.statusValue;
        sortValue = searchFields.sortValue;
        orderValue = searchFields.orderValue;
      }

      const token = getToken();
      const server = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;
      const getUrl = `http://${server}/plan/search/${textValue}/${statusValue}/` +
        `${sortValue}/${orderValue}/${cursor.primary}/${cursor.secondary}/?accessToken=${token}`;
      let obj = {};

      // get our search results
      const results = await fetch(getUrl);
      setLoading(false);
      if (results.ok) {

        // if this is a new search clear all of the previous results
        obj = await results.json();
        if (cursor.primary === "null") {
          setPlans([...obj.plans]);
        } else {
          setPlans([...plans, ...obj.plans]);
        }
        setCursor(obj.nextCursor);

      } else {
        // we got a bad status code. Show the error
        obj = await results.json();
        setErrorMessage(obj.error);
        if (results.status === 500) {
          setErrorMessage("An internal server error occurred. Please try again later.");
        }
        if (results.status === 404) {
          setPlans([]);
        }
      }
    } catch (err) {
      // show error message if error while searching
      setErrorMessage("An internal server error occurred. Please try again later.");
    }
  }

  if (!pageError) {
    return (
      <div css={style}>
        <PageSpinner loading={loading} />
        <NavBar />
        <div id="advisor-home-container">
          <div id="advisor-home-contents-container">

            {recentPlans.length ? (
              <RecentPlans recentPlans={recentPlans}/>
            ) : (
              null
            )}

            <FindPlans onSearch={cursor => searchPlans(cursor)}/>

            <div className="home-error-message-container">{errorMessage}</div>

            {plans.length ? (
              <SearchResults plans={plans} cursor={cursor}
                onLoadMore={cursor => searchPlans(cursor)} />
            ) : (
              null
            )}

          </div>
        </div>
      </div>
    );
  } else {
    return <PageInternalError />;
  }
}
export default AdvisorHome;

AdvisorHome.propTypes = {
  history: PropTypes.object
};