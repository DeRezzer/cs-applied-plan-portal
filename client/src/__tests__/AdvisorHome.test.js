/* eslint-disable no-undef */
import React from "react";
import {MemoryRouter} from "react-router-dom";
import renderer from "react-test-renderer";

import AdvisorHome from "../components/advisor_home/AdvisorHome";


describe("Advisor home", () => {

  it("renders successfully", () => {
    const wrapper = renderer.create(
      <MemoryRouter>
        <AdvisorHome />
      </MemoryRouter>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});