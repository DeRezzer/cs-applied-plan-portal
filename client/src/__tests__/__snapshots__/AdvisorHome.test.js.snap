// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advisor home renders successfully 1`] = `
<div
  className="css-1nx3o09-AdvisorHome"
  id="advisor-home-page"
>
  <div
    className="navbar-parent css-12g91iz-Navbar"
    id="navbar"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <p
        className="osu-logo"
      >
        OSU CS Applied Plan Portal
      </p>
    </a>
    <div
      className="right-container"
    >
      <div
        className="notification-dropdown css-3i4jcj-Notifications"
      >
        <button
          className="drop-button-notification"
          data-count={0}
        >
          Notifications
          <span
            className="badge"
          />
          <i
            className="fa fa-caret-down"
          />
        </button>
        <div
          className="dropdown-content"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            <p>
              No new notifications
            </p>
          </a>
        </div>
      </div>
      <button
        className="logout-button css-1k1xu2y-Logout"
        onClick={[Function]}
      >
        Log Out
      </button>
    </div>
  </div>
  <div
    id="advisor-home-container"
  >
    <div
      id="advisor-home-contents-container"
    >
      <div
        className="css-1ewmpfq-FindPlans"
        id="plan-search-container"
      >
        <h2>
          Search Plans
        </h2>
        <form
          id="search-form"
          onSubmit={[Function]}
        >
          <input
            id="input-search"
            type="text"
          />
          <button
            id="search-plan-button"
          >
            Search
          </button>
        </form>
        <div
          id="filter-container"
        >
          <select
            className="advisor-plan-filter"
            defaultValue="5"
            id="select-status"
          >
            <option
              value="5"
            >
              Any Status
            </option>
            <option
              value="2"
            >
              Awaiting Review
            </option>
            <option
              value="3"
            >
              Awaiting Final Review
            </option>
            <option
              value="1"
            >
              Awaiting Student Changes
            </option>
            <option
              value="4"
            >
              Accepted
            </option>
            <option
              value="0"
            >
              Rejected
            </option>
          </select>
        </div>
      </div>
      <div
        className="css-1mu153g-SearchResults"
      >
        <div
          className="prompt-container"
        >
          <h3>
            Search for plans...
          </h3>
        </div>
      </div>
    </div>
  </div>
</div>
`;
